mqtt.go

```diff
@@ -19,598 +24,620 @@ const (
+		// unanswered. These block, so a size of 1 is sufficient.
+		packIDToChan: make(map[uint16]mqttPacketChan, 1),
+		// Channel to the mqtt server. This can possibly get backed up on a slow
+		// network. The size of 8 is arbitrary. If it is full, we will block
+		// until a timeout (specified by the delegate) and return an error.
+		outPktCh: make(mqttPacketChan, 8),
+		// Connection requests are blocked, so 1 is sufficient
+		connRespCh: make(mqttPacketChan, 1),
+		// These are created by the delegate, so it's up to them to create
+		// channels of sufficient size or we will fail.
+		pingRespCh: pingAckCh,
+		queueAckCh: queueAckCh,
 	}
 }
 
-func (mc *mqttConn) sendEvent(e *DeviceEvent) error {
-	var qos byte
+func (conn *mqttConn) Close() {
+	conn.closeAllChannels()
+}
 
-	switch e.qos {
-	case QOSAtMostOnce:
-		qos = packet.QOSAtMostOnce
-	case QOSAtLeastOnce:
-		qos = packet.QOSAtLeastOnce
+// We only queue pings (PINGREQ) and publishes (PUBLISH). Theoretically, we
+// could queue subscribes (SUBSCRIBE) since they have packet ID's like
+// publishes. But, to be able to handle queueing and sending for the same type
+// of packet would make the code overly complicated.
+func (conn *mqttConn) QueuePacket(p packet.Packet,
+	timeout time.Duration) (uint16, error) {
+
+	var packetID uint16
+	switch p.Type() {
+	case packet.PINGREQ:
+		// We send these as is, without adding a packet ID
+	case packet.PUBLISH:
+		pubPkt := p.(*packet.PublishPacket)
+		packetID = conn.getPacketID()
+		pubPkt.PacketID = uint16(packetID)
 	default:
-		return errors.New("unsupported qos level")
+		logError("[MQTT] Unhandled packet type: %s", p.Type())
+		return 0, fmt.Errorf("Unhandled packet type for queueing")
 	}
 
-	payload, _ := json.Marshal(e)
-
-	p := packet.NewPublishPacket()
-	p.PacketID = mc.getPacketID()
-	p.Message = packet.Message{
-		Topic:   fmt.Sprintf("/devices/%d/event", mc.dc.DeviceID),
-		QOS:     qos,
-		Payload: payload,
-	}
-
-	mc.outPacket <- p
-
-	logInfo("[MQTT] event delivery for packet ID %d", p.PacketID)
-
-	if e.qos == QOSAtMostOnce {
-		return nil
-	}
-
-	logInfo("[MQTT] waiting for PUBACK for packet ID %d", p.PacketID)
-
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
 	select {
-	case <-time.After(deviceEventRetryInterval):
-		msg := fmt.Sprintf("did not receive PUBACK of event delivery for packet ID %d", p.PacketID)
-		logError("[MQTT] %s", msg)
-		mc.reportError(fmt.Errorf("event delivery timeout: %s", msg))
-
-	case ack := <-mc.puback:
-		if ack.PacketID == p.PacketID {
-			logInfo("[MQTT] received PUBACK for packet ID %d", ack.PacketID)
-			return nil
-		}
-
-		msg := fmt.Sprintf("sendEvent() received packet that does not match the Ack packet ID (%d) and the expected packet ID (%d).",
-			ack.PacketID, p.PacketID)
-		logInfo("[MQTT] %s", msg)
-		mc.reportError(fmt.Errorf("event delivery unmatched Ack Packet: %s", msg))
+	case conn.outPktCh <- p:
```

Comment:
If QoS level is one then client user should be wait PUBACK by self using delegator?

Reply:
The client will wait for the PUBACK by the packet ID. The client should wait for the PUBACK.